### PR TITLE
machines: Configure  Network model type

### PR DESCRIPTION
**Pull request by KKoukiou**
* [x] some preparation work: PR #10251
* [x] Need this for adding the inactiveXML to global state PR #10120

### CODE REVIEW

**Comment by garrett** [comment]
From what I gather, here are the reasons why someone would use different cards:

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/virtualization_administration_guide/sect-virtualization-troubleshooting-kvm_networking_performance

- virtio has the best performance
  - it also is the same as e1000, right? if so, then should it be listed twice?
- e1000 and e1000e would be chosen for compatibility?
  - according to the above URL, virtio == e1000?
  - vmware supports e1000/e1000e — is this a reason why one would pick this card to emulate?
  - e1000e is a newer / faster card to emulate?
- rtl8139 would also be chosen for compatibility reasons?

**Comment by garrett** [comment]
Thanks for helping me get it working over IRC!

Here's a screenshot of the machines page on my computer:

screenshot_2018-09-26 virtual machines - sunny

There's another link in the row for the device source (virbr0), and clicking it takes me to the network page. When clicking the model type, nothing happens for me.

**Comment by garrett** [comment]
We talked about a few things in IRC:

- indicate a restart is needed
   - most likely with an icon that has a tooltip, possibly pficon-pending, but there are a few other suitable ones
   - this should also show up in a few other places (like vCPU), but this is outside the scope of this specific PR
- inline edit, like an embedded dropdown

I need to make some mockups. :grin:

**Comment by Antique** [comment]
> From what I gather, here are the reasons why someone would use different cards:
> 
> https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/virtualization_administration_guide/sect-virtualization-troubleshooting-kvm_networking_performance
> 
> * virtio has the best performance
>   
>   * it also is the same as e1000, right? (according to the above URL, virtio == e1000) if so, then should it be listed twice?

This is not correct, virtio != e1000.  The e1000 model is real card that is emulated by QEMU, the same applies to e1000e and rtl8139. Virtio is virtual card and has the best performance because it doesn't need to emulate how real card behaves, the only drawback is that you need to have driver for virtio network card in the guest. Most linux guests have that driver but windows doesn't and you need to install the driver in order to use virtio network.

> * e1000 and e1000e would be chosen for compatibility?

Correct, e1000e is fairly new HW so it might not be supported by older OSes.

>   * vmware supports e1000/e1000e — is this a reason why one would pick this card to emulate?

The best choice is virtio if you can install driver for virtio into the guest OS, otherwise you should pick the newest emulated card that is supported by the guest OS.

>   * e1000e is a newer / faster card to emulate?

I'm not sure whether it's faster then e1000, the main difference is that e1000 is legacy PCI device but e1000e is moder PCIe device.

> * rtl8139 would also be chosen for compatibility reasons?

Yes, for some really old OSes.

**Comment by Antique** [comment]
One thing about the design, instead of having multiple windows for every single value that we can change for network device can we have only one window where you would be able to change all/some values?

For example, the network-source depends on which network-type is selected, so in order to change network-type you would have to also select new network-source.

So there could be some type of "edit" button at the end of the row which would open editing window and there you would be able to change the values.

**Comment by garrett** [comment]
I've taken all the feedback above (including my own) and produced two mockups we can discuss.

From here, we can either choose one of the two and go with it, build upon one of the two with a few modifications, or go back to the drawing board.

1. The first approach uses **inline editing** and simply lets us change the model type.
   - I don't have source editing here, however. If it would be editable, it would change based on the model type. The two would be interdependent as @Antique mentions above.
   - (The source is still a link to the networks page, which I think is a bit odd.)

2. The second set of mockups uses a **dialog** with an edit button.
   - The network source changes when the model type changes, like @Antique wrote.

Both mockups have the concept of restarts being needed. The message appears when the new type != currently used type. (In other words: The UI always shows the selected type and shows the message when necessary.)

|Inline mockups|Dialog mockups|
|---|---|
|vm-network-edit-inline|vm-network-edit-dialog|

**Comment by garrett** [comment]
Oh, if an unplug/replug is good enough for network changes, then we could actually forgo the restart messaging in some cases and do save/apply. I suppose it would/could cause issues with current network connections, however.

**Comment by Antique** [comment]
@garrett Thanks a lot for the summary :) . I'm glad that the final agreement is to use dialog, I personally prefer that option and as you've wrote we can reuse it for adding new network interfaces.

To the "Model type" name I would change it to "model" only, we can safely remove the type.  "chipset" feels wrong since it's the whole card, "card" could be possible replacement but in that case it would be nice to add some "tooltip" to mention that it actually means "model" in domain XML.  Or we can use virt-manager name for it "Device model".

The restart option is tricky and I don't think that there will be simple way how to automatically figure out whether it's safe to restart it or not compared to the NFS example so I would use only [ Cancel ] [ Apply ] buttons with the warning message if restart of a guest is required.  We also need to somehow document that the QEMU process must be restarted, reboot from guest is not enough, so the restart means that you need to fully shutdown the guest and start it again in order to have the changes applied.

I would not implement the change operation as unplug & plug operation because it disconnect the guest from network and it might interrupt some operation running in the guest or mess-up the connection.

**So to summarize it:**

1. implement dialog to change network interfaces with [ Cancel ] and [ Apply ] buttons only
2. possibly rename "Model type" to "Model" or "Device model" (I would avoid "Chipset" or "Card" if possible) (probably different pull request)
3. possibly rename existing plug/unplug button to connect/disconnect, if you are working with virtualization for some time it might be misleading that it plugs/unplugs the whole device, but it actually plugs/unplugs the imaginary cable to/from the virtual card.
4. implement create new device/delete existing device with hot-plug/hot-unplug functionality (new pull request)

**Comment by garrett** [comment]
> [ Cancel ] and [ Apply ] buttons

1. I made it [Cancel] and [Save], as it isn't applied until restarting the VM.

2. "Model" works for me (already changed it in the mockup before I saw your comment).  :+1:

3. I'm not a fan of plug/unplug, but I think connect/disconnect might be problematic, as connect/disconnect is often used for connecting to a network. Enable/Disable might work? Or we could keep it plug/unplug, possibly with some tooltip text if people don't know what it means.

4. Yeah, create new would be great. And, yeah, it should be another issue+PR.  :+1:

**Comment by garrett** [comment]
We _probably_ need to show the entire card state in the dialog. Possibly something like:


vm-network-edit-dialog-preview

**Comment by garrett** [comment]
Thought process behind more complicated dialog:

1. if we just have "edit" for the row and only a couple of fields to edit, it's not really clear
2. it's not obvious what one is then editing, unless we have additional info (the dialog will cover identifiers like a mac address and the network type)

At least it could be re-used for an add network device dialog.

---

If not everything can be edited or if we do not want to inflate the scope much, then we could make some of the fields non-editable, like this pruned-down example:

vm-network-edit-dialog-preview-readonly

**Comment by garrett** [comment]
Here's a version that drops target and changes additional settings to MTU:

vm-network-edit-dialog-preview-mtu

**Comment by garrett** [comment]
@KKoukiou: I think we're on the same page already, based on the conversation here and on IRC. Thanks again for all your work on this too!

Here's a revised complete flow with all the changes, so it's documented for others to follow along (and in case something above might not be clear):

vm-network-edit-dialog

Please note that I also changed the MAC address (switching from "coffee coffee" to "coffee decaf" :coffee:), so it also gets a notification icon. I figured I might as well represent > 1 thing being changed, and also keep the other 2 the same.

After looking at it here, I see the VM isn't running... so it's silly to have a reboot warning. Please pretend it is running. :grin:

**Comment by garrett** [comment]
Including currently modified mockups, for reference. (There's probably nothing new or surprising here. It's just good to keep it in sync.)

vm-network-edit-dialog

**Comment by garrett** [comment]
It's coming along nicely!

There are still a couple of interaction issues, I think:

1. If the type dropdown needs a source, it doesn't show up unless you re-select the tyle.
2. The restart on change should only show up after the form has been modified.
3. I had the modification icon show up next to each item. It's probably fine as-is. (This is debatable — and looking at it in action, I think it's fine as-is just being in the bottom left of the dialog.)
4. I should probably add some more comments to the CSS as to _why_ we're doing things the way we're doing it. :wink:

**Comment by KKoukiou** [comment]
> It's coming along nicely!
> 
> There are still a couple of interaction issues, I think:
> 
>     1. If the type dropdown needs a source, it doesn't show up unless you re-select the tyle.

I can't reproduce this, maybe we have different setup, I really can't find what might cause this.
> 
>     2. The restart on change should only show up after the form has been modified.

Fixed.

> 
>     3. I had the modification icon show up next to each item. It's probably fine as-is. (This is debatable — and looking at it in action, I think it's fine as-is just being in the bottom left of the dialog.)

Yes, I 'd rather leave it as is as well :) 
> 
>     4. I should probably add some more comments to the CSS as to _why_ we're doing things the way we're doing it. wink

**Comment by garrett** [comment]
I removed unused CSS @ https://github.com/cockpit-project/cockpit/pull/9949/commits/6e1b050a97d9b7b9c3de8eb431b3b865b99da98d & added a couple comments @ https://github.com/cockpit-project/cockpit/pull/9949/commits/e86488c492fafa5d5b7653191d38202b994b0d16, then squashed back into the previous commit and did a force-push into this PR.

**Comment by garrett** [comment]
Regarding issue 1:

Here's the dialog by default:

screen shot 2018-10-15 at 11 50 03-fullpage

And after selecting the same network source, the source dropdown shows up:

screen shot 2018-10-15 at 11 50 11-fullpage

I have no virtual networks, so perhaps that's causing the issue here? (Just a wild guess.)

**Comment by KKoukiou** [comment]
> Regarding issue 1:
> 
> Here's the dialog by default:
> 
> screen shot 2018-10-15 at 11 50 03-fullpage
> 
> And after selecting the same network source, the source dropdown shows up:
> 
> screen shot 2018-10-15 at 11 50 11-fullpage
> 
> I have no virtual networks, so perhaps that's causing the issue here? (Just a wild guess.)

Ah exactly that's why, fixing it.

**Comment by KKoukiou** [comment]
This [failure](https://fedorapeople.org/groups/cockpit/logs/pull-9949-20181015-140505-6fc1f2f8-verify-debian-testing/log.html#245) on debian testing is definitely related.  Needs investigation.

**Comment by KKoukiou** [comment]
@martinpitt  this is ready for review. The testCreate is the only failure from machines tests, but it's totally unrelated to this change.

File: pkg/machines/components/vmnetworktab.jsx [hunk 1]
```diff
@@ -110,7 +157,7 @@ const VmNetworkTab = function ({ vm, dispatch, hostDevices }) {
         <div className="machines-network-list">
             {message}
             <Listing columnTitles={detailMap.map(target => target.name)} actions={null} emptyCaption=''>
-                {vm.interfaces.sort().map(target => {
+                {vm.interfaces.map(target => {
```
Comment (mvollmer) [inline pkg/machines/components/vmnetworktab.jsx]: Why is this not sorted anymore?

**Comment by mvollmer** [comment]
The pending icon looks like this for me:
net-pending

Is there some CSS missing?

**Comment by mvollmer** [comment]
The message says "Changes will apply on VM restart" and there is a "Restart" button, but clicking that button doesn't actually apply the changes.  One needs to either "Force restart" or "Shutdown" the VM in order for the changes to apply.  (That's probably because "Restart" does not actually terminate the qemu process but just kicks the guest into its bootloader.)

File: pkg/machines/components/vmnetworktab.jsx [hunk 2]
```diff
@@ -23,12 +23,30 @@ import { changeNetworkState } from "../actions/provider-actions.es6";
...
 
-const VmNetworkTab = function ({ vm, dispatch, hostDevices }) {
+const VmNetworkTab = function ({ vm, dispatch, config, hostDevices, networks }) {
     const id = vmId(vm.name);
 
+    const warningInactive = (id) => {
+        return (
+            <Tooltip tip={_("Changes will apply on VM restart")} pos='top'>
+                <span id={id} className='fa-stack fa-sm'>
+                    <i className='fa fa-circle-thin fa-stack-2x fa-stack-xs' />
```
Comment (mvollmer) [inline pkg/machines/components/vmnetworktab.jsx]: Shouldn't this also be `pf-pending`?

File: pkg/machines/components/vmnetworktab.jsx (hunk 1 above)
Comment (KKoukiou) [reply pkg/machines/components/vmnetworktab.jsx]: My mistake. Reverting it back.

File: pkg/machines/components/vmnetworktab.jsx (hunk 2 above)
Comment (KKoukiou) [reply pkg/machines/components/vmnetworktab.jsx]: Right, as the other one, I forgot to change this one. Thanks

**Comment by KKoukiou** [comment]
> The message says "Changes will apply on VM restart" and there is a "Restart" button, but clicking that button doesn't actually apply the changes. One needs to either "Force restart" or "Shutdown" the VM in order for the changes to apply. (That's probably because "Restart" does not actually terminate the qemu process but just kicks the guest into its bootloader.)
> 
> So that is confusing. Can we say "Changes will apply on VM shutdown"?

That's exactly the reason why the changes are not applied. I think your point makes sense. I'll change that.

**Comment by KKoukiou** [comment]
> The pending icon looks like this for me:
> net-pending
> 
> Is there some CSS missing?

Yes, fixed that.

**Comment by KKoukiou** [comment]
Nic edit button:
nic-edit

Nic edit modal dialog:
edit-nic-modal-dialog

Nic edit modal dialog warning message:
edit-nic-modal-dailog-model-changed

Nic edit warning icon tooltip:
nic-edit-changes-pending-tooltip

**Comment by martinpitt** [comment]
The fedora-29 test run points out an error:
```
> error: Warning: Failed prop type: The prop `networks` is marked as required in `Vm`, but its value is `undefined`.
    in Vm (created by HostVmsList)
    in HostVmsList (created by HostVmsListDecorated)
    in div (created by HostVmsListDecorated)
    in HostVmsListDecorated (created by App)
    in div (created by App)
    in App
```

**Comment by KKoukiou** [comment]
> The fedora-29 test run points out an error:
> 
> ```
> > error: Warning: Failed prop type: The prop `networks` is marked as required in `Vm`, but its value is `undefined`.
>     in Vm (created by HostVmsList)
>     in HostVmsList (created by HostVmsListDecorated)
>     in div (created by HostVmsListDecorated)
>     in HostVmsListDecorated (created by App)
>     in div (created by App)
>     in App
> ```
First time I see this :/
Looks like a race condition that shows up if the networks object is created in the store later than the Vms component is mounted. Trying the following diff.
```bash
diff --git a/pkg/machines/libvirt-dbus.es6 b/pkg/machines/libvirt-dbus.es6
index 97c580d1a..96c8b8d9a 100644
--- a/pkg/machines/libvirt-dbus.es6
+++ b/pkg/machines/libvirt-dbus.es6
@@ -406,10 +406,10 @@ LIBVIRT_DBUS_PROVIDER = {
                 dispatch(checkLibvirtStatus(libvirtServiceName));
                 dbus_client(connectionName);
                 startEventMonitor(dispatch, connectionName, libvirtServiceName);
-                doGetAllVms(dispatch, connectionName);
                 dispatch(getStoragePools(connectionName));
                 dispatch(getNetworks(connectionName));
                 dispatch(getHypervisorMaxVCPU(connectionName));
+                doGetAllVms(dispatch, connectionName);
             };
         }

```

**Comment by martinpitt** [comment]
Still the [same error](https://fedorapeople.org/groups/cockpit/logs/pull-9949-20181024-070303-18e0ebdc-verify-fedora-29/log.html#243)

**Comment by KKoukiou** [comment]
Fix for the ovirt test:
``` bash
diff --git a/pkg/ovirt/components/App.jsx b/pkg/ovirt/components/App.jsx
index 3454f44e4..e022117ac 100644
--- a/pkg/ovirt/components/App.jsx
+++ b/pkg/ovirt/components/App.jsx
@@ -135,6 +135,7 @@ const HostVmsListDecorated = ({ vms, config, systemInfo, ui, dispatch, host }) =
                          ui={ui}
                          storagePools={{}}
                          dispatch={dispatch}
+                         networks={{}}
                          actions={actions} />
         </div>
     );

```

**Comment by KKoukiou** [comment]
@martinpitt @mvollmer ovirt test was fixed with the above diff. All other failures are unrelated.
